### PR TITLE
Search: Fix a bug when searching folders under the general folder

### PR DESCRIPTION
**Pull request by lingyufei**
**What is this feature?**

This PR is to fix a bug that will return all matched nested folders when searching folders only under the general folder.

In Grafana v11, I create a few nested folders like the below image: 
image

When using the search API to search the folders named 'b' under the general folder, it returns all the folders named 'b', including the folder `b` under the folder `a`. But it should only return one folder under the general folder.
`http://localhost:3000/api/search?query=b&dash=dash-folder&folderUIDs=general`. The result:

```
[
    {
        "id": 2021,
        "uid": "ddswszjpzzoxse",
        "title": "b",
        "uri": "db/b",
        "url": "/dashboards/f/ddswszjpzzoxse/b",
        "slug": "",
        "type": "dash-folder",
        "tags": [],
        "isStarred": false,
        "folderUid": "ddswszbukighsc",
        "folderTitle": "a",
        "sortMeta": 0
    },
    {
        "id": 2022,
        "uid": "adswt078w2dc0d",
        "title": "b",
        "uri": "db/b",
        "url": "/dashboards/f/adswt078w2dc0d/b",
        "slug": "",
        "type": "dash-folder",
        "tags": [],
        "isStarred": false,
        "sortMeta": 0
    }
]
```

I fixed this issue by using the SQL `dashboard.folder_uid is NULL ` when `NestedFoldersEnabled` is enabled in `FolderUIDFilter`, rather than using `dashboard.folder_id = 0`

Please see the `dashboard` table below. When `NestedFoldersEnabled` is enabled, the folder_uid of a nested folder is not null, but its folder_id is still 0.
Thus, it will return all matched nested folders if we use `dashboard.folder_id = 0` to match folders under the general folder.

image

**Why do we need this feature?**

To fix a bug that will return all matched nested folders when searching folders under the general folder.

**Who is this feature for?**

All users using Search API

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by lingyufei** [comment]
@joshhunt  Please kindly check if this change looks good to you. Thanks.

**Comment by lingyufei** [comment]
@inf0rmer  @joshhunt  Please kindly review this PR. Thanks!

**Comment by lingyufei** [comment]
@inf0rmer @joshhunt @itsmylife Please kindly review this PR. Thanks!

**Comment by itsmylife** [comment]
I am tagging the squad @grafana/grafana-search-and-storage who owns that part of the code for a review.

**Comment by lingyufei** [comment]
> I am tagging the squad @grafana/grafana-search-and-storage who owns that part of the code for a review.

@itsmylife Thanks. It has been a long time since I submitted this PR, and I haven't received any feedback for now.

It's an important bug fix/enhancement because I'm using this Search Function daily but it's malfunctioning.

**Comment by suntala** [comment]
@lingyufei thank you for your PR! I will be reviewing next week.

**Review by suntala — COMMENTED** [review]
The change looks good so far. It looks like we will need some tests though and I am afraid that means expanding the existing test to check `f.NestedFoldersEnabled`. 

Just so you are aware, it is on our roadmap to remove nested folders/subfolders code since the feature is now generally available. Testing would be simpler after that refactor. But if you wanted to wait for that, it might take a while :)

**Comment by lingyufei** [comment]
@suntala Thanks for your suggestion. I can add some tests for this change currently. 
Do you mean adding tests to check f.NestedFoldersEnabled to the test file `filter.test`?

**Comment by suntala** [comment]
> Maybe I can add some tests to check the returned SQL when f.NestedFoldersEnabled is enabled to the test file filter.test?

Yes, assuming you mean [filters_test.go](https://github.com/grafana/grafana/blob/e93337f40a311e8154aac81fe685735e3abddd1c/pkg/services/sqlstore/searchstore/filters_test.go), that's how I would approach it as well.

**Comment by lingyufei** [comment]
@suntala I have added a unit test `TestFolderUIDFilterWithNestedFoldersEnabled `. The cases are the same with `TestFolderUIDFilter`, only the excepted results differ.

Please check if this looks good to you.

**Comment by suntala** [comment]
Thank you for adding the tests. I'll take a closer look tomorrow so you might want to hold off on making changes until then. At first glance, it seems to me we could combine the test cases you added with the existing one in `filters_test.go`.

**Comment by lingyufei** [comment]
@suntala Sure. Let me know if you have any ideas about this unit test.

File: pkg/services/sqlstore/searchstore/filters_test.go
Comment (suntala) [inline pkg/services/sqlstore/searchstore/filters_test.go:61]: We could add these test cases to the test above. The way to do so would be to expand the `testCases` struct to track whether `NestedFoldersEnabled` should be set to `true` or `false`. This new field in `testCases` could also be called "nestedFoldersEnabled".

So instead of setting `NestedFoldersEnabled` to `true` on line 102, we would set it to  `tc.nestedFoldersEnabled`. The test cases above would have `tc.nestedFoldersEnabled` set to `false` while the new ones you added would have it set to `true`.

**Comment by lingyufei** [comment]
@suntala Thanks for your good advise. This can make our test code graceful.

I have updated the test code based on your advise. Please check if it looks good to you.

**Comment by joshhunt** [comment]
@lingyufei I'm glad this is being looked at. Just to clarify - you ran into this through your own use of the Search API? I don't believe Grafana's UI calls `/api/search?folderUIDs=general` to search folders anywhere?

For API consistency, searching for folders should function the same as searching for dashboards. Specifying a folderUID should return results for only that folder, and not any descendants.

For the product, it's exceptionally important that if the `folderUIDs` is not specified then all dashboards (or folders) are searched for. This still lets the global search on the root `/dashboards` Browse Dashboards search work.

Glancing over this PR, I think this is all the correct behaviour we want 👍 @suntala

**Comment by lingyufei** [comment]
> @lingyufei I'm glad this is being looked at. Just to clarify - you ran into this through your own use of the Search API? I don't believe Grafana's UI calls `/api/search?folderUIDs=general` to search folders anywhere?
> 
> For API consistency, searching for folders should function the same as searching for dashboards. Specifying a folderUID should return results for only that folder, and not any descendants.
> 
> For the product, it's exceptionally important that if the `folderUIDs` is not specified then all dashboards (or folders) are searched for. This still lets the global search on the root `/dashboards` Browse Dashboards search work.
> 
> Glancing over this PR, I think this is all the correct behaviour we want 👍 @suntala

@joshhunt Yes. I'm using this Grafana Search API to search folders/dashboards from my other project. I found this issue when trying to search folders in the root/general folder via this API.

**Comment by suntala** [comment]
At first glance it looks good. Thank you for making the refactor. I'll take one more look tomorrow, but unless I missed something, you can expect an approval then. I'll also take a look at which checks need enabling before the PR can get merged.

**Review by suntala — APPROVED** [review]
Thank you for this!

**Comment by suntala** [comment]
@lingyufei the build for the continuous-integration/drone/pr is failing. Looks like you will need to rebase your branch on a newer version of `main` and force push it.

**Comment by lingyufei** [comment]
@suntala I have rebased my branch on the latest main. Let's try to go through the ci/cd again! 
All tests were passed locally.

**Comment by lingyufei** [comment]
@suntala Thanks for your patience. 🙂

**Pull request merged**
